File: Miscellaneous/Python/206. Reverse Linked List.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseList(self, head):
        curr = head
        prev = None
        
        while curr is not None:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        head = prev
        
        return head
        
obj = Solution()
list = ListNode(1)
list.next = ListNode(2)
list.next.next = ListNode(3)
list.next.next.next = ListNode(4)
ans = obj.reverseList(list)
while ans is not None:
    print(ans.val)
    ans = ans.next

File: Miscellaneous/Golang/206. Reverse Linked List.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {

	list := &ListNode{1, nil}
	list.Next = &ListNode{2, nil}
	list.Next.Next = &ListNode{3, nil}
	list.Next.Next.Next = &ListNode{4, nil}
	result := reverseList(list)

	for result != nil {
		fmt.Print(result.Val, " ")
		result = result.Next
	}

}

func reverseList(head *ListNode) *ListNode {
	curr := head
	var prev *ListNode

	for curr != nil {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}
	head = prev

	return head
}
